app/Services/DavClient/SynchronizeAddressBook.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace App\Services\DavClient;
+
+use App\Models\User\User;
+use Illuminate\Support\Arr;
+use App\Services\BaseService;
+use App\Helpers\AccountHelper;
+use App\Models\Account\Account;
+use Illuminate\Support\Facades\Log;
+use GuzzleHttp\Client as GuzzleClient;
+use GuzzleHttp\Exception\ClientException;
+use App\Models\Account\AddressBookSubscription;
+use App\Services\DavClient\Utils\Dav\DavClient;
+use App\Services\DavClient\Utils\Model\SyncDto;
+use App\Services\DavClient\Utils\AddressBookSynchronizer;
+use App\Http\Controllers\DAV\Backend\CardDAV\CardDAVBackend;
+
+class SynchronizeAddressBook extends BaseService
+{
+    /**
+     * Get the validation rules that apply to the service.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'account_id' => 'required|integer|exists:accounts,id',
+            'user_id' => 'required|integer|exists:users,id',
+            'addressbook_subscription_id' => 'required|integer|exists:addressbook_subscriptions,id',
+            'force' => 'nullable|boolean',
+        ];
+    }
+
+    /**
+     * @param  array  $data
+     * @return void
+     */
+    public function execute(array $data, GuzzleClient $httpClient = null)
+    {
+        $this->validate($data);
+
+        $account = Account::find($data['account_id']);
+        if (AccountHelper::hasReachedContactLimit($account)
```

Comment:
If `$account` doesn't exist above, I don’t think this will work.

Shouldn't we do `findOrFail` instead?

Reply:
actually, it's certain the account exist already, thanks to the `validate` 2 lines below.